enterprise/backend/src/metabase_enterprise/caching/strategies.clj

```diff
@@ -0,0 +1,204 @@
+                            [[:inline 1] :ordering]]
+                   :where  [:and [:= :model "database"]   [:= :model_id (:database_id card)]]}
+                  {:from   [:cache_config]
+                   :select [:id
+                            [[:inline 1] :ordering]]
+                   :where  [:= :model "root"]}]
+          q      {:select   [:id]
+                  :limit    1
+                  :from     [[{:union-all qs} :unused_alias]]
+                  :order-by :ordering}
+          config (t2/select-one :model/CacheConfig :id q)]
+      (if config
+        (merge {:type       (:strategy config)
+                :updated_at (:updated_at config)
+                :payload    (:payload config)}
+               (:config config))
+        (when-let [ttl (granular-ttl card dashboard-id)]
+          {:type     :duration
+           :duration ttl
+           :unit     "hours"})))))
+
+;;; Strategy execution
+
+(defmulti fetch-cache-stmt-ee*
+  "Generate prepared statement for a db cache backend for a given strategy"
+  (fn [strategy _hash _conn] (:type strategy)))
+
+(defmethod fetch-cache-stmt-ee* :ttl [strategy query-hash conn]
+  (backend.db/fetch-cache-stmt-ttl strategy query-hash conn))
+
+(defmethod fetch-cache-stmt-ee* :duration [strategy query-hash conn]
+  (assert (:duration strategy))
+  (assert (:unit strategy))
+  (let [duration  (t/duration (:duration strategy) (keyword (:unit strategy)))
+        timestamp (t/minus (t/offset-date-time) duration)]
+    (backend.db/prepare-statement conn query-hash timestamp)))
+
+;; NOTE: something changes `:updated_at` to `:updated-at`
+(defmethod fetch-cache-stmt-ee* :schedule [{:keys [updated-at] :as strategy} query-hash conn]
+  (if-not updated-at
+    (log/debugf "Caching strategy %s was never run yet" (pr-str strategy))
+    (backend.db/prepare-statement conn query-hash updated-at)))
+
+(defmethod fetch-cache-stmt-ee* :query [{:keys [updated-at] :as strategy} query-hash conn]
+  (if-not updated-at
+    (log/debugf "Caching strategy %s was never run yet" (pr-str strategy))
+    (backend.db/prepare-statement conn query-hash updated-at)))
+
+(defmethod fetch-cache-stmt-ee* :nocache [_ _ _]
+  nil)
+
+(defenterprise fetch-cache-stmt
+  "Returns prepared statement to query for db cache backend."
+  :feature :cache-granular-controls
+  [strategy query-hash conn]
+  (fetch-cache-stmt-ee* strategy query-hash conn))
+
+;;; Jobs
+
+(defn- refresh-schedule-configs []
+  (let [now (t/offset-date-time)]
+    (doseq [item (t2/select :model/CacheConfig :strategy :schedule {:where [:or
+                                                                            [:= :next_run_at nil]
+                                                                            [:<= :next_run_at [:now]]]})
+            :let [cron     (-> (cron/cron-schedule (-> item :config :schedule))
+                               cron/finalize)
+                  ;; needed by the tests, or the cron will use it's own current date
```

Comment:
I'm not sure what this means. Do the tests stub out the `java-date` time and this syncs the `cron` with that virtual time? Without this line what time does it use?

Reply:
Is there not a way we can do this less intrusively?

Reply:
So if we don't set a date, then the quartz lib will set "current date" without using our date mocks. So if we want this to behave with our tests - we have to the "current date" ourselves. 

I'm happy to rephrase, though, if you have any suggestions.

Reply:
How about "Synchronize cron runner to our app time, which may be mocked in tests"?